Modules/_struct.c

```diff
@@ -1492,8 +1521,10 @@ Struct___init___impl(PyStructObject *self, PyObject *format)
 
     Py_SETREF(self->s_format, format);
 
-    ret = prepare_s(self);
-    return ret;
+    if (prepare_s(self) < 0) {
```

Comment:
It looks as though we're duplicating work at this point: `prepare_s` is being called twice for a new `struct.Struct` instance. I think we can simply drop most of the content of the `__init__` method: no need to do the `prepare_s`, the `Py_SETREF` or the type check.

Removing those things would technically be a backwards incompatible change, since it means you can't deliberately call `__init__` on an existing `struct.Struct` instance with a new format string (or at least, that that `__init__` call wouldn't have any useful effect), but I think that's not unreasonable, especially if this change only goes into 3.12.

Reply:
Fixed.